modules/041-linstor/hooks/metrics.go

```diff
@@ -0,0 +1,303 @@
+					"resourcedefinitions.internal.linstor.linbit.com",
+					"resources.internal.linstor.linbit.com"},
+			},
+			FilterFunc: applyCRDVersionsFilter,
+		},
+	},
+}
+
+var _ = sdk.RegisterFunc(linstorMetricsHookConfig, collectLinstorMetrics)
+
+func applyCRDVersionsFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var s apiextensions.CustomResourceDefinition
+	err := sdk.FromUnstructured(obj, &s)
+	if err != nil {
+		return "", err
+	}
+
+	for _, v := range s.Spec.Versions {
+		if v.Storage {
+			return v.Name, nil
+		}
+	}
+	return "", fmt.Errorf("Can not find storage version for %s", s.Name)
+}
+
+func applyLinstorNodeFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var s linstorObj
+	err := sdk.FromUnstructured(obj, &s)
+	if err != nil {
+		return "", err
+	}
+
+	nodeType, ok := s.Spec["node_type"].(int64)
+	if !ok {
+		nodeTypeFloat64, _ := s.Spec["node_type"].(float64)
+		nodeType = int64(nodeTypeFloat64)
+	}
+
+	return linstorNodeSnapshot{
+		Type: nodeType,
+	}, nil
+}
+
+func applyLinstorSPFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var s linstorObj
+	err := sdk.FromUnstructured(obj, &s)
+	if err != nil {
+		return "", err
+	}
+
+	driverName, _ := s.Spec["driver_name"].(string)
+
+	return linstorSPSnapshot{
+		Driver: driverName,
+	}, nil
+}
+
+func applyLinstorRDFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var s linstorObj
+	err := sdk.FromUnstructured(obj, &s)
+	if err != nil {
+		return "", err
+	}
+
+	snapshotName, _ := s.Spec["snapshot_name"].(string)
+
+	return linstorRDSnapshot{
+		IsSnapshot: snapshotName != "",
+	}, nil
+}
+
+func applyLinstorResourceFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var s linstorObj
+	err := sdk.FromUnstructured(obj, &s)
+	if err != nil {
+		return "", err
+	}
+
+	snapshotName, _ := s.Spec["snapshot_name"].(string)
+
+	return linstorResourceSnapshot{
+		IsSnapshot: snapshotName != "",
+	}, nil
+}
+
+// collectLinstorMetrics
+//
+// synopsis:
+//   Waits for internal LINSTOR resources be installed then
+//   collects general statistics from them
+//
+func collectLinstorMetrics(input *go_hook.HookInput) error {
+	input.MetricsCollector.Set(telemetry.WrapName("linstor_enabled"), 1, map[string]string{})
+
+	// LINSTOR manages it's own internal CRDs, so we need to wait for them before starting the watch
+	if linstorMetricsHookConfig.Kubernetes[0].Kind == "" {
+
+		var version string
+		for _, sRaw := range input.Snapshots[linstorCRDsSnapshot] {
+			discoveredVersion := sRaw.(string)
+			if version == "" {
+				version = discoveredVersion
+			} else if version != discoveredVersion {
+				return fmt.Errorf("LINSTOR internal CRDs have different storage versions")
+			}
+		}
+
+		if len(input.Snapshots[linstorCRDsSnapshot]) >= 4 {
+			// LINSTOR installed
+			input.LogEntry.Infof("LINSTOR internal CRDs installed, update kind for binding linstor resources to collect metrics")
```

Comment:
Maybe `Debugf`?

Reply:
Thus we would need to change it in other modules too:

https://github.com/deckhouse/deckhouse/blob/4320c132b23d2cbad74d66c022104857583ad720/modules/040-node-manager/hooks/get_crds.go#L173
https://github.com/deckhouse/deckhouse/blob/4320c132b23d2cbad74d66c022104857583ad720/modules/490-virtualization/hooks/disk_handler.go#L196
https://github.com/deckhouse/deckhouse/blob/4320c132b23d2cbad74d66c022104857583ad720/modules/490-virtualization/hooks/vm_handler.go#L139
https://github.com/deckhouse/deckhouse/blob/4320c132b23d2cbad74d66c022104857583ad720/modules/490-virtualization/hooks/create_registry_secret_for_vmi.go#L162

Reply:
ok, stay it